File: matlab/newton.m
format long

f=input('Ingrese la función f(x), por ejemplo: @(x) exp(-x)-x  :');
x0=input('Ingrese el valor inicial x0:  ');
n=input ('Ingrese el número de iteraciones máximo:  ');
tol=input ('Ingrese la tolerancia:  ');

[x,e,i] = newtonraphson(f,x0,n,tol);
function [x0,er,i] = newtonraphson(f,x0,n,tol)
syms x;
inicial=x0;
sp = '%.10f';
spa='%.10f %.10f %.10f %.10f %.10f\n';
disp (['i ' 'x ' 'fx ' 'df' 'error abs ']);
fx=double(subs(f,x0));
df=matlabFunction( diff(f(x)));
dfx=double(subs(df,x0));
i=0;
er=tol+1;
while fx~=0 && er>tol && i<n && dfx~=0
    if i==0
        er=nan;
    end
    resp=[i double(x0) double(fx) double(dfx) double(er)];
    fprintf( spa, resp.' ) ;
    x1=x0-(fx/dfx);
    fx=double(subs(f,x1));
    dfx=double(subs(df,x1));
    er=abs(x1-x0);
    x0=x1;
    i=i+1;
end
resp=[i double(x0) double(fx) double(dfx) double(er)];
fprintf( spa, resp.' ) ;
if fx==0
    disp([num2str(x0,sp) ' es raíz.']);
elseif er<tol
    disp([num2str(x1,sp) ' es aproximación a una raíz con tolerancia ' num2str(tol,sp)]);
elseif dfx==0
    disp([num2str(x1,sp) 'es una posible raíz multiple']);
else
    disp(['El método fracasó con ' num2str(i) ' iteraciones.']);
end
ejex = linspace(inicial, x0+100*tol);
figure
plot(ejex, double(subs(f,ejex)))
grid
end

File: python/newton.py
def newton(f, df, x0, tol, n):
    """
    Calculates the root of an equation by Newton method
    Parameters:
            f: Function f(x)
           df: Derivative of function f(x)
           x0: Initial guess
          tol: Tolerance
     iter_max: Maximum number of iterations
    Outpus:
         root: Root value
         iter: Used iterations
    converged: Found the root
    """
    fx= f(x0)
    dfx= df(x0)
    er = tol + 1
    i=0


    while fx!=0 and er>tol and i<n and dfx!=0:
        x1 = x0 - (fx/dfx)
        fx= f(x1)
        dfx= df(x1)
        er= abs(x1 - x0)
        x0=x1
        i=i+1

        print("i:{:03d} x: {:.10f} fx: {:.10f} error abs: {:.10f} \n"
              .format(i, x0, fx, er ), end="")

        if fx== 0:
            print(x0, "Es raíz")
        elif er <= tol:
            print(x1, ' es aproximación a una raíz con tolerancia ', tol)
        elif dfx == 0:
            print(x1, ' es una posible raíz multiple')
        else:
            print('El método fracasó con ', i, ' iteraciones.')


